### PR TITLE
Extender API for Profile folder creation with meta-file

### PR DESCRIPTION
**Pull request by JillieBeanSim**
## Extender API for Profile folder creation with meta-file
issue: https://github.com/zowe/vscode-extension-for-zowe/issues/1254

Extender API that can be utilized by Zowe Explorer Extenders to have profile folder and meta file created upon initialization if not already existing.
FTP implementation of using this Extender API to no longer have a pre-requisite of the FTP CLI plug-in

ZE-zFTP-initialize

## Release Notes

Milestone: 1.15.0

Changelog: 
1.  FTP no longer has a pre-requisite on the `@zowe/zos-ftp-for-zowe-cli`
2.  Zowe Explorer extenders can now have their profile type's folder with meta file created in the `/.zowe/profiles` home directory upon initialization by calling the `ZoweExplorerApiRegister.getExplorerExtenderApi().initForZowe(type: string, meta:imperative.ICommandProfileTypeConfiguration[])` during their activation with Zowe Explorer.

## Types of changes

What types of changes does your code introduce to Zowe Explorer?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Updates to Documentation or Tests (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This checklist will be used as reference for both the contributor and the reviewer_

- [ ] I have read the [CONTRIBUTOR GUIDANCE](https://github.com/zowe/vscode-extension-for-zowe/wiki/Best-Practices:-Contributor-Guidance) wiki
- [ ] PR title follows [Conventional Commits Guidelines](https://www.conventionalcommits.org/en/v1.0.0-beta.2/)
- [ ] PR Description is included
- [ ] gif or screenshot is included if visual changes are made
- [ ] `yarn workspace vscode-extension-for-zowe vscode:prepublish` has been executed
- [ ] All checks have passed (DCO, Jenkins and Code Coverage)
- [ ] I have added unit test and it is passing
- [ ] I have added integration test and it is passing
- [ ] There is coverage for the code that I have added
- [ ] I have tested it manually and there are no regressions found
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any PR dependencies have been merged and published (if appropriate)

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by JillieBeanSim** [comment]
@tiantn @std4lqi would you mind checking this out as you did a lot of work on the FTP extension?  I did have to make some changes to the file ZoweExplorerFtpJesApi due to errors that showed up after updating to the @zowe/zos-ftp-for-zowe-cli 1.5.0.  Please make sure I didn't break anything there, thanks.

**Comment by zFernand0** [comment]
I hope `FTP no longer has a pre-requisite on the @zowe/zos-ftp-for-zowe-cli` doesn't cause confusion.
The extension still depends on the plugin's SDK/APIs, it's just not a prereq for users to install the CLI and the plugin 😋

**Comment by JillieBeanSim** [comment]
> I hope `FTP no longer has a pre-requisite on the @zowe/zos-ftp-for-zowe-cli` doesn't cause confusion.
> The extension still depends on the plugin's SDK/APIs, it's just not a prereq for users to install the CLI and the plugin 😋

Let me reword that lol

**Comment by JillieBeanSim** [comment]
FYI a lot of the files changed are due to running `yarn run pretty` due to formatting issues in the master branch and threw errors when I tried to run `yarn run package`

**Review by lauren-li — APPROVED** [review]
I think this PR looks good! I uninstalled the FTP CLI plugin, removed my `zftp` profiles folder, and then tested this PR using its generated VSIX files for Zowe Explorer and the FTP extension. The `zftp` profiles folder gets created again with the appropriate `meta` YAML file inside, and I am able to then create working zftp profiles in Zowe Explorer. All unit and expected integration tests are passing.

Regarding the extra whitespace changes, I would usually suggest to do this in a separate PR, but since it seems this PR requires these changes to pass the build tests, maybe it's okay to keep them in the PR this time.
- Note: It seems like we keep running into the issue with these whitespace changes, so after these changes are merged in, I agree that we could add `yarn run pretty` somewhere in the PR checks so it doesn't keep happening.

Thanks @JillieBeanSim for your hard work on this PR!

**Comment by tiantn** [comment]
> @tiantn @std4lqi would you mind checking this out as you did a lot of work on the FTP extension? I did have to make some changes to the file ZoweExplorerFtpJesApi due to errors that showed up after updating to the @zowe/zos-ftp-for-zowe-cli 1.5.0. Please make sure I didn't break anything there, thanks.

Hi @JillieBeanSim, I  have a try with it and the FTP extension works well. 
It seems when the VS Code starts the zftp file folder will be created so the VS Code FTP extension no longer need to install zftp plugin to get the pre-requisite(@zowe/cli and @zowe/imperative). Is that right? Thank you!

**Comment by JillieBeanSim** [comment]
> > @tiantn @std4lqi would you mind checking this out as you did a lot of work on the FTP extension? I did have to make some changes to the file ZoweExplorerFtpJesApi due to errors that showed up after updating to the @zowe/zos-ftp-for-zowe-cli 1.5.0. Please make sure I didn't break anything there, thanks.
> 
> Hi @JillieBeanSim, I have a try with it and the FTP extension works well.
> It seems when the VS Code starts the zftp file folder will be created so the VS Code FTP extension no longer need to install zftp plugin to get the pre-requisite(@zowe/cli and @zowe/imperative). Is that right? Thank you!

@tiantn  Thanks for checking this out and testing it.  That is correct, once this PR is merged and published @zowe/cli and @zowe/zos-ftp-for-zowe-cli will no longer be an installation prerequisite for the FTP VS Code extension.  Users will be able to install Zowe Explorer and the FTP extension into VS Code and start working right away with nothing else required to create FTP profiles.

**Comment by tiantn** [comment]
> > > @tiantn @std4lqi would you mind checking this out as you did a lot of work on the FTP extension? I did have to make some changes to the file ZoweExplorerFtpJesApi due to errors that showed up after updating to the @zowe/zos-ftp-for-zowe-cli 1.5.0. Please make sure I didn't break anything there, thanks.
> > 
> > 
> > Hi @JillieBeanSim, I have a try with it and the FTP extension works well.
> > It seems when the VS Code starts the zftp file folder will be created so the VS Code FTP extension no longer need to install zftp plugin to get the pre-requisite(@zowe/cli and @zowe/imperative). Is that right? Thank you!
> 
> @tiantn Thanks for checking this out and testing it. That is correct, once this PR is merged and published @zowe/cli and @zowe/zos-ftp-for-zowe-cli will no longer be an installation prerequisite for the FTP VS Code extension. Users will be able to install Zowe Explorer and the FTP extension into VS Code and start working right away with nothing else required to create FTP profiles.

Thank you！@JillieBeanSim 😊

**Comment by crawr** [comment]
@IgorCATech note to update the FTP doc to remove: FTP implementation of using this Extender API to no longer have a pre-requisite of the FTP CLI plug-in

**Comment by jellypuno** [comment]
@JillieBeanSim I am experiencing this error:

image

To reproduce:
1. delete the .zowe folder and let Zowe Explorer create it during initialization
2. Create a profile. 
         - If I create a zFtp then the config error happens.
         - If I create a zosmf profile, it is success but with the loadAll error message
         
       Maybe I am missing something or I have a wrong setup?

**Comment by JillieBeanSim** [comment]
@jellypuno did you build, package, and install Zowe Explorer and FTP extension from this branch?

**Comment by JillieBeanSim** [comment]
ZE-FTP

**Comment by jellypuno** [comment]
@JillieBeanSim Aha! I didn't do that. I just run debug. Ok I will install it

**Comment by JillieBeanSim** [comment]
> @JillieBeanSim Aha! I didn't do that. I just run debug. Ok I will install it

Me too usually but this PR has code changes across all 3 pieces of the mono-repo.

**Comment by jellypuno** [comment]
I forgot to install Zowe Explorer 😄 let me try again

**Comment by jellypuno** [comment]
@JillieBeanSim I was successful! 🎉 but I found an error in update profile.

I created a profile with wrong info then I updated it with the right info (all in ZE). It says updated successfully. After that I updated it again but it is showing me the wrong info that I have from before. I checked the yaml file and the right info is there.

image

**Comment by JillieBeanSim** [comment]
@jellypuno I am not seeing this happen on marketplace or this branch versions.  I even went in and cleaned my AppData/Roaming/Code/{Cache, CachedData, CachedExtensions, CachedExtensionsVSIXs, and Code Cache} folders locally to make sure I didn't have anything cached to keep me from seeing what you experienced.

**Comment by jellypuno** [comment]
@JillieBeanSim ok. let me clear my cache as well

**Comment by jellypuno** [comment]
@JillieBeanSim The clearing of cache worked for me. I am  not experiencing the issue anymore

**Review by jellypuno — APPROVED** [review]
